### PR TITLE
feat: paginated vaults PE-7081

### PR DESCRIPTION
**Pull request by fedellen**

### CODE REVIEW

File: tests/e2e/cjs/index.test.js [hunk 1]
```diff
@@ -339,6 +339,33 @@ describe('e2e cjs tests', async () => {
...
+        for (const [vaultId, vault] of Object.entries(vaultsForAddress)) {
+          assert(typeof vaultId === 'string');
+          assert(typeof vault === 'object');
+          assert(typeof vault.balance === 'number');
+          assert(typeof vault.startTimestamp === 'number');
+          assert(typeof vault.endTimestamp === 'number');
+        }
+      });
+    });
   });
```
Comment (dtfiedler) [inline tests/e2e/cjs/index.test.js:369]: going forward - don't worry about updating these tests. they are more to validate imports work as expected and the full test suite can just live in `esm/index.test.js)

**Comment by fedellen** [comment]
added!

File: README.md [hunk 2]
```diff
@@ -396,6 +397,51 @@ const balances = await io.getBalances({
...
+        "vaultIdOne": {
+          "balance": 1000000,
+          "startTimestamp": 123,
+          "endTimestamp": 4567
+        },
+        "vaultIdTwo": {
+          "balance": 1000000,
+          "startTimestamp": 123,
+          "endTimestamp": 4567
+        }
```
[the remark this answers is not in the document]
Comment (fedellen) [reply README.md]: updated

File: README.md [hunk 3]
```diff
@@ -396,6 +397,50 @@ const balances = await io.getBalances({
 
 </details>
 
+#### `getVaults({ cursor, limit, sortBy, sortOrder })`
+
+Retrieves the vaults of the IO process, paginated and sorted by the specified criteria. The `cursor` used for pagination is the last wallet address from the previous request.
```
Comment (arielmelendez) [inline README.md:402]: Consider: `Retrieves all locked-balance user vaults of the IO process...` to help disambiguate from delegation withdraw vaults.

File: README.md [hunk 4]
```diff
@@ -396,6 +397,50 @@ const balances = await io.getBalances({
...
+      "address": "QGWqtJdLLgm2ehFWiiPzMaoFLD50CnGuzZIPEdoDRGQ",
+      "vaultId": "vaultIdTwo",
+      "balance": 1000000,
+      "startTimestamp": 123,
+      "endTimestamp": 4567
+    }
+    // ...98 other addresses with vaults
+  ],
+  "hasMore": true,
+  "nextCursor": "1",
```
Comment (arielmelendez) [inline README.md:437]: Nit: could have a better nextCursor example since they're wallet addresses in this context

File: README.md (hunk 3 above)
Comment (fedellen) [reply README.md]: f2ec83c

File: README.md (hunk 4 above)
Comment (fedellen) [reply README.md]: f2ec83c

File: src/types/io.ts [hunk 5]
```diff
@@ -191,11 +191,16 @@ export type AoGatewayServices =
...
+  address: WalletAddress;
+  vaultId: string;
+};
+
 export type AoGateway = {
   settings: AoGatewaySettings;
   stats: AoGatewayStats;
   totalDelegatedStake: number;
-  vaults: Record<WalletAddress, AoVaultData>;
+  vaults: AoWalletVault;
```
Comment (dtfiedler) [inline src/types/io.ts:203]: this is is till the previous types. AoWalletVault is only returned by getVault/getVaults API atm

File: tests/e2e/esm/index.test.js [hunk 6]
```diff
@@ -431,6 +431,57 @@ describe('e2e esm tests', async () => {
         assert(io instanceof IOWriteable);
       }
     });
+
+    // TODO: Make a vault within this test environment's context to cover this
```
Comment (dtfiedler) [inline tests/e2e/esm/index.test.js:435]: this is a larger pending action item. what we can do is create a stubbed vault on devnet to validate the behavior against

File: tests/e2e/esm/index.test.js [hunk 7]
```diff
@@ -431,6 +431,57 @@ describe('e2e esm tests', async () => {
...
+      vaults.items.forEach(
+        ({ address, vaultId, balance, endTimestamp, startTimestamp }) => {
+          assert(typeof address === 'string');
+          assert(typeof balance === 'number');
+          assert(typeof startTimestamp === 'number');
+          assert(typeof endTimestamp === 'number');
+          assert(typeof vaultId === 'string');
+        },
+      );
+    });
```
Comment (dtfiedler) [inline tests/e2e/esm/index.test.js:484]: nit: add one more test with custom sort parameters

File: src/types/io.ts (hunk 5 above)
Comment (fedellen) [reply src/types/io.ts]: 3f281ed

Comment (fedellen) [reply src/types/io.ts]: vaultId is not be represented in that API?

File: tests/e2e/esm/index.test.js (hunk 7 above)
Comment (fedellen) [reply tests/e2e/esm/index.test.js:484]: 8b2593a

**Review by dtfiedler — APPROVED** [review]

**Comment by dtfiedler** [comment]
looks like the esm integration test with custom sort failed

**Comment by fedellen** [comment]
> looks like the esm integration test with custom sort failed

re-ran the action with updated devnet process and now it passed 🙌

**Pull request merged**